app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java

```diff
@@ -53,9 +58,31 @@ public EnrollmentRepository(@NonNull Context context,
     public Flowable<List<FieldViewModel>> list() {
 
         return d2.enrollmentModule().enrollments().uid(enrollmentUid).get()
-                .flatMap(enrollment -> d2.programModule().programTrackedEntityAttributes().withRenderType().byProgram().eq(enrollment.program()).get()).toFlowable()
-                .flatMapIterable(programTrackedEntityAttributes -> programTrackedEntityAttributes)
-                .map(this::transform).toList().toFlowable();
+                .flatMap(enrollment -> d2.programModule().programs().uid(enrollment.program()).get())
+                .flatMap(program -> d2.programModule().programSections().byProgramUid().eq(program.uid()).withAttributes().get()
+                        .flatMap(programSections -> {
+                            if (programSections.isEmpty()) {
+                                return d2.programModule().programTrackedEntityAttributes().withRenderType().byProgram().eq(program.uid()).get().toFlowable()
+                                        .flatMapIterable(programTrackedEntityAttributes -> programTrackedEntityAttributes)
+                                        .map(this::transform).toList();
+                            } else {
+                                List<FieldViewModel> fields = new ArrayList<>();
+                                for (ProgramSection section : programSections) {
+                                    fields.add(transformSection(section));
+                                    for (TrackedEntityAttribute attribute : section.attributes()) {
+                                        ProgramTrackedEntityAttribute programTrackedEntityAttribute =
+                                                d2.programModule().programTrackedEntityAttributes()
+                                                        .withRenderType()
+                                                        .byProgram().eq(program.uid())
+                                                        .byTrackedEntityAttribute().eq(attribute.uid())
+                                                        .one().blockingGet();
+                                        fields.add(transform(programTrackedEntityAttribute));
+                                    }
+                                }
+                                return Single.just(fields);
+                            }
+                        })
+                ).toFlowable();
```

Comment:
could you extract some of this logic to other method/class so we can add unit test?

Reply:
or add unit test to this method in the class?

Reply:
thanks!

Reply:
I'll create two public methods with the VisisbleForTesting tag

Reply:
I think you don't need it and you can tests the whole method?

Reply:
visible for testing is not usually a good practise, I use it for only exceptions..